File: recursion_i/principle_of_recursion/reverse_string/reverse_string.cpp
/**
 * Author: Nathan Bockisch
 * Date: October 20th, 2021
 **/

class Solution {
public:
    void recursiveReverser(vector<char>& s, int start, int end) {
        // Base case
        if (end <= start) return;
        
        char tmp = s[start];
        s[start] = s[end];
        s[end] = tmp;
        
        // Swap the ith and size - ith characters
        recursiveReverser(s, ++start, --end);
    }
    
    void reverseString(vector<char>& s) {
        recursiveReverser(s, 0, s.size() - 1);
    }
};

File: array_and_string/two_pointer_technique/reverse_string/reverse_string.rs
/*
 * Author: Nathan Bockisch
 * Date: July 12, 2022
 */
impl Solution {
    pub fn reverse_string(s: &mut Vec<char>) {
        let (mut l_ptr, mut r_ptr) = (0, s.len() - 1);

        while (r_ptr > l_ptr) {
            s.swap(l_ptr, r_ptr);
            l_ptr += 1;
            r_ptr -= 1;
        }
    }
}
